platform/linux-generic/include/odp_ipsec_internal.h

```diff
@@ -84,8 +84,20 @@ int _odp_ipsec_status_send(odp_queue_t queue,
 
 #define IPSEC_MAX_SALT_LEN	4    /**< Maximum salt length in bytes */
 
-/* 32 is minimum required by the standard. We do not support more */
-#define IPSEC_ANTIREPLAY_WS	32
+/* The minimum supported AR window size */
+#define IPSEC_AR_WIN_SIZE_MIN	32
+
+/* The maximum supported AR window size */
+#define IPSEC_AR_WIN_SIZE_MAX	4096
+
+/* For a 64-bit bucket size */
+#define IPSEC_AR_WIN_BUCKET_BITS	6
+#define IPSEC_AR_WIN_BUCKET_SIZE	(1 << IPSEC_AR_WIN_BUCKET_BITS)
+#define IPSEC_AR_WIN_BITLOC_MASK	(IPSEC_AR_WIN_BUCKET_SIZE - 1)
+
+/* Maximum number of buckets */
+#define IPSEC_AR_WIN_BUCKET_MAX		\
+	(IPSEC_AR_WIN_SIZE_MAX / IPSEC_AR_WIN_BUCKET_SIZE + 1)
```

Comment:
This assumes that IPSEC_AR_WIN_SIZE_MAX is divisible by IPSEC_AR_WIN_BUCKET_SIZE and can break if someone later does not notice it and changes the max window size.

I would propose defining the following macro and using it to determine the number of buckets for both the maximum number of buckets and the actual number of buckets used for an SA.

```
/*
 * We need one extra bucket in addition to the buckets that contain
 * part of the window.
 */
#define IPSEC_AR_WIN_NUM_BUCKETS(window_size) \
        (((window_size) - 1) / IPSEC_AR_WIN_BUCKET_SIZE + 2)
```

Reply:
The macro "IPSEC_AR_WIN_BUCKET_MAX" is defined for bucket array size initialization "uint64_t bucket_arr[IPSEC_AR_WIN_BUCKET_MAX];". So I will define it as 

`#define IPSEC_AR_WIN_BUCKET_MAX    65`

Reply:
> So I will define it as #define IPSEC_AR_WIN_BUCKET_MAX 65

Why would you want to do that instead of letting the macro calculate the number of buckets? The magic number 65 would have to be changed when the max window size is changes and getting it wrong may open up a security problem in the implementation. Since the correct number of buckets can be, and needs to be, calculated from the window size, it is better to let the computer, i.e. the macro, do the calculation and thus also make the calculation explicit, instead of hiding it behind the magic number and counting on that the next person changing the code figures it all out.

So, please define IPSEC_AR_WIN_BUCKET_MAX in terms of IPSEC_AR_WIN_SIZE_MAX and IPSEC_AR_WIN_BUCKET_SIZE which it depends on and not through a magic number that comes out of the sky.

Reply:
Understood it now,  visualized your first comment a bit differently, sorry for that.
Now, the newly proposed macro is used in both the cases as per your first comment.

1) The maximum number of buckets macro is defined as.

```
#define IPSEC_AR_WIN_BUCKET_MAX         \
        IPSEC_AR_WIN_NUM_BUCKETS(IPSEC_AR_WIN_SIZE_MAX)
```

2) The actual number of buckets used for a given window size is determined as

`num_bkts = IPSEC_AR_WIN_NUM_BUCKETS(ipsec_sa->in.ar.win_size);`